src/main/scala/org/broadinstitute/dsde/workbench/leonardo/auth/IamProxyAuthProvider.scala

```diff
@@ -0,0 +1,98 @@
+package org.broadinstitute.dsde.workbench.leonardo.auth
+
+import cats.implicits._
+import akka.actor.ActorSystem
+import com.typesafe.config.Config
+import net.ceedubs.ficus.Ficus._
+import org.broadinstitute.dsde.workbench.google.GoogleCredentialModes.Token
+import org.broadinstitute.dsde.workbench.google.{GoogleIamDAO, HttpGoogleIamDAO}
+import org.broadinstitute.dsde.workbench.leonardo.model.NotebookClusterActions.NotebookClusterAction
+import org.broadinstitute.dsde.workbench.leonardo.model.ProjectActions.ProjectAction
+import org.broadinstitute.dsde.workbench.leonardo.model.google.ClusterName
+import org.broadinstitute.dsde.workbench.leonardo.model.{LeoAuthProvider, ServiceAccountProvider}
+import org.broadinstitute.dsde.workbench.model.{UserInfo, WorkbenchEmail}
+import org.broadinstitute.dsde.workbench.model.google.{GoogleProject, IamPermission}
+
+import scala.concurrent.{ExecutionContext, Future}
+
+class IamProxyAuthProvider(config: Config, serviceAccountProvider: ServiceAccountProvider)(implicit val system: ActorSystem) extends LeoAuthProvider(config, serviceAccountProvider) {
+
+  val applicationName: String = config.getValue("applicationName").toString()
+  val requiredPermissions: Set[IamPermission] = config.as[Set[String]]("requiredProjectIamPermissions").map(p => IamPermission(p))
+
+  protected def petGoogleIamDao(token: String)(implicit executionContext: ExecutionContext): GoogleIamDAO = {
+    new HttpGoogleIamDAO(applicationName, Token(() => token), "google")
+  }
+
+  protected def checkUserAccess(userInfo: UserInfo, googleProject: GoogleProject)(implicit executionContext: ExecutionContext): Future[Boolean] = {
+    val iamDAO: GoogleIamDAO = petGoogleIamDao(userInfo.accessToken.token)
+    val foundPermissions = iamDAO.testIamPermission(googleProject, requiredPermissions)
+    Future.successful(requiredPermissions == foundPermissions)
```

Comment:
Doesn't look right -- you're comparing a `Future[Set[IamPermission]]` with a `Set[IamPermission]`. Probably do something like this instead:
```
iamDAO.testIamPermission(googleProject, requiredPermissions).map { foundPermissions =>
   foundPermissions == requiredPermissions
}
```